hbase-server/src/main/java/org/apache/hadoop/hbase/wal/AbstractRecoveredEditsOutputSink.java

```diff
@@ -0,0 +1,283 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.wal;
+
+import static org.apache.hadoop.hbase.wal.WALSplitUtil.getCompletedRecoveredEditsFilePath;
+import static org.apache.hadoop.hbase.wal.WALSplitUtil.getRegionSplitEditsPath;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.CellUtil;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.log.HBaseMarkers;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.hadoop.hbase.util.FSUtils;
+import org.apache.hadoop.ipc.RemoteException;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hbase.thirdparty.org.apache.commons.collections4.MapUtils;
+
+@InterfaceAudience.Private
+public abstract class AbstractRecoveredEditsOutputSink extends OutputSink {
+  private static final Logger LOG = LoggerFactory.getLogger(RecoveredEditsOutputSink.class);
+  private WALSplitter walSplitter;
+  private FileSystem walFS;
+  private Configuration conf;
+  private final ConcurrentMap<String, Long> regionMaximumEditLogSeqNum = new ConcurrentHashMap<>();
+
+  public AbstractRecoveredEditsOutputSink(WALSplitter walSplitter,
+      WALSplitter.PipelineController controller, EntryBuffers entryBuffers, int numWriters) {
+    super(controller, entryBuffers, numWriters);
+    this.walSplitter = walSplitter;
+    this.walFS = walSplitter.walFS;
+    this.conf = walSplitter.conf;
+  }
+
+  /**
+   * @return a writer that wraps a {@link WALProvider.Writer} and its Path. Caller should close.
+   */
+  protected RecoveredEditsWriter createWriter(TableName tableName, byte[] region, long seqId)
+      throws IOException {
+    // If we already decided that this region doesn't get any output
+    // we don't need to check again.
+    if (blacklistedRegions.contains(region)) {
+      return null;
+    }
+    String tmpDirName = walSplitter.conf
+        .get(HConstants.TEMPORARY_FS_DIRECTORY_KEY, HConstants.DEFAULT_TEMPORARY_HDFS_DIRECTORY);
+    Path regionEditsPath = getRegionSplitEditsPath(tableName, region, seqId,
+        walSplitter.getFileBeingSplit().getPath().getName(), tmpDirName, conf);
+    if (regionEditsPath == null) {
+      blacklistedRegions.add(region);
+      return null;
+    }
```

Comment:
This logic belongs elsewhere? In WALSplitter?

Reply:
I thought it should belong here. It is a part of createRecoveredEditsWriter.